native_renderer/plugin.cpp

```diff
@@ -0,0 +1,477 @@
+#include <cassert>
```

Comment:
The file has barely any comments. please add detailed comments to what each function/code chunk is doing